RFC: constants in patterns (#3535)
rust-lang/rfcs: text/0000-constants-in-patterns.md

  Furthermore, `+0.0` (written just `0.0` in Rust) and `-0.0` have different bit representations, but they *do* compare equal.
  We can easily declare NaNs to not have structural equality, and reject them in patterns, since there is no situation where it makes sense to have a NaN in a pattern.
  However, for zeroes this is more tricky -- allowing matching on `1.0` but rejecting `0.0` is likely going to be considered extremely arbitrary and inconsistent.
  This RFC therefore suggests that all zeroes should have structural equality, and match all zeroes.
  `0.0` and `-0.0` as a value of a constant used in a pattern will all match both `0.0` and `-0.0`, as they do today.
  (As a literal, `0.0` and `-0.0` are currently both accepted but `+0.0` is not. This RFC does not propose to change anything here, though we could consider linting against `-0.0`.)

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

When a constant `C` of type `T` is used as a pattern, we first check that `T: PartialEq`.
Furthermore we require that the value of `C` *has (recursive) structural equality*, which is defined recursively as follows:

> - Integers, `bool` and `char` values, and references always have structural equality.

Reviewer: I think this isn't quite correct: just because a value of type `&Foo` is a reference, doesn't mean it has structural equality. the pointed-to value also needs to have structural equality.

Author: *oops* you are right of course...

    ```rust
    enum E { Var1, Var2 }

    const BEST_VAR: E = E::Var1;

    fn is_best(e: E) -> bool { matches!(e, BEST_VAR) }
    fn is_var1(e: E) -> bool { matches!(e, E::Var1) }
    ```

    It would be very surprising if those two functions could behave differently.
  It follows that we cannot allow `PartialEq` to affect the behavior of constants in patterns.

- We do not want to expose equality tests on types where the library author did not explicitly expose an equality test.
  This means not allowing matching on constants whose type does not implement `PartialEq`.
  It also means not allowing matching on constants where running `PartialEq` would behave different from the corresponding pattern:
  the pattern can be accessing private fields, and the `PartialEq` implementation provided by the library might be treating those fields in a particular way;
  we should not let people write patterns that "bypass" any such treatment and do structural matching when the crate author does not provide a structural `PartialEq`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Constants can be used as patterns, but only if their type implements `PartialEq`.

> Morevoer, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields (recursively):

Reviewer 2: ```suggestion
Moreover, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields (recursively):
```

equality of this type is defined fully by the equality on its fields.

For matching on values of enum type, it is sufficient if the actually chosen
variant has structural equality; other variants do not matter:

```rust
struct NonStructuralEq(i32);

impl PartialEq for NonStructuralEq {
    fn eq(&self) -> bool { true }
}

#[derive(PartialEq)]
enum MyEnum { GoodVariant(i32), NonStructuralVariant(NonStructuralEq) }

// This constant *can* be used in a pattern.
const C: MyEnum = MyEnum::GoodVariant(0);
```

Overall we say that the *value* of the constant must have recursive structural equality,
which is the case when all the types that actually appear recursively in the type (ignoring "other" enum variants) have structural equality.

Most of the values of primitive Rust types have structural equality (integers, `bool`, `char`, references), but two families of types need special consideration:

- Pointer types (raw pointers and function pointers): these compare by test the memory address for equality.
  It is unclear whether that should be considered "structural", but it is fairly clear that this should be considered a bad idea:

>   Rust makes basically no guarantees for when two function pointers are equal or inequal

Reviewer 2: One could argue this spelling is valid, but it jumped out as unusual to me.

```suggestion
  Rust makes basically no guarantees for when two function pointers are equal or unequal
```

## Breaking changes

This RFC breaks code that compiles today, but only code that already emits a future compatibility lint:
- Matching on constants that do not implement `PartialEq` sometimes accidentally works, but triggers `const_patterns_without_partial_eq`.
  This lint landed with Rust 1.74 (the most recent stable release), and is shown in dependencies as well.
- Matching on `struct`/`enum` that do not `derive(PartialEq)` is accidentally possible under some conditions, but triggers `indirect_structural_match`.
  This has been a future-compatibility lint for many years, though it is currently not shown in dependencies.
- Matching on function pointers, or raw pointers that are not defined as a constant integer, triggers `pointer_structural_match`.
  This only recently landed (Rust 1.75, currently in beta), and is not currently shown in dependencies.
  Crater found [three cases](https://github.com/rust-lang/rust/pull/116930#issuecomment-1784648989) across the ecosystem where `match` was used to compare function pointers;
  that code is buggy for the reasons mentioned above that make comparing function pointers unreliable.

> - Matching on floats triggers illegal_floating_point_literal_pattern`. This triggers on *all* float matches, not juts the forbidden ones.

Reviewer 2: Typos

```suggestion
- Matching on floats triggers `illegal_floating_point_literal_pattern`. This triggers on *all* float matches, not just the forbidden ones.
```

# Unresolved questions
[unresolved-questions]: #unresolved-questions

None at this time.

# Future possibilities
[future-possibilities]: #future-possibilities

- At some point, we might want to stabilize the `StructuralPartialEq` trait.
  There are however plenty of open questions here:
  + Should `StructuralPartialEq` be an unsafe trait?
    That trait has a clear semantic meaning, so making it `unsafe` to be able to rely on it for soundness is appealing.
    In particular, we could then be *sure* that pattern matching always has the same semantics as `==`.
    However, that would mean `forbid(unsafe_code)` crates could not `derive(PartialEq)` any more, which is clearly not acceptable.
    As a consequence of `StructuralPartialEq` being safe, it is actually possible to write patterns that behave different from the `PartialEq` that is explicitly defined and intentionally exposed on that type,
    but only if one of the involved crates implements `StructuralPartialEq` incorrectly.
    This can lead to semver issues and logic bugs, but that is all allowed for safe traits.
    However, this also means unsafe code cannot rely on `==` and pattern matching having the same behavior.

>     One could consider doing some special magic to `derive(PartialEq)` to remain compatible with `forbid(unsafe_code)`, but that still seems suboptimal.

Reviewer 3: Such magic already exists, it's called `#[allow_internal_unsafe]`.
It is currently used on definitions of `format_args!` and `thread_local!`.

Author: Oh I didn't know about that -- I'll mention it in the RFC, thanks.

  we should not let people write patterns that "bypass" any such treatment and do structural matching when the crate author does not provide a structural `PartialEq`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Constants can be used as patterns, but only if their type implements `PartialEq`.
Moreover, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields (recursively):

```rust
#[derive(PartialEq)] // code fails to build if we remove this or replace it by a manual impl
enum E { Var1, Var2 }

const BEST_VAR: E = E::VAR1;

fn is_best(e: E) -> bool { matches!(e, BEST_VAR) }
```

```rust
#[derive(PartialEq)]
struct S { f1: i32, f2: E }

const MY_S: S = S { f1: 42, f2: BEST_VAR };

// Removing *either* `derive` or implementing `PartialEq` manually would lead to rejecting this code
fn is_mine(s: S) -> bool { matches!(s, MY_S) }
```

We say that a type that derives `PartialEq` has "structural equality": the
equality of this type is defined fully by the equality on its fields.

For matching on values of enum type, it is sufficient if the actually chosen

> variant has structural equality; other variants do not matter:

Author: Not sure if it's unfortunate, people certainly [like and use it](https://github.com/rust-lang/rust/issues/62411#issuecomment-510604193).

  It has been around for years, but is not currently shown in dependencies.

When the RFC gets accepted, the floating-point lint should be adjusted to only cover the cases we are really going to reject,
and all of them should be shown in dependencies or directly turned into hard errors.

## Compiler/library cleanup

There also exists the `nontrivial_structural_match` future compatibility lint;
it is not needed for this RFC so it can be removed when the RFC gets accepted.

Similarly, the `StructuralEq` trait no longer serves a purpose and can be removed.

# Drawbacks
[drawbacks]: #drawbacks

- The biggest drawback of this proposal is that it conflates `derive(PartialEq)` with a semver-stable promise that this type will always have structural equality.
  Once a type has `derive(PartialEq)`, it may appear in patterns, so replacing this `PartialEq` by a custom implementation is a breaking change.
  Once the `StructuralPartialEq` trait is stable, `derive(PartialEq)` *can* be replaced by a custom implementation as long as one also implements `StructuralPartialEq`,

>   but that entails a promise that the `impl` still behaves structurally, including on all private fields.

Reviewer 4: (It would be breaking, which means we almost certainly shouldn't do it, but) Given the logic that a `const` pattern is “logically desugared” into the corresponding literal pattern, it could be justified in forbidding matching over any type with inaccessible fields (or `#[non_exhaustive]`).

(Fun: because this sees into private fields, you can exhaustively match a type with all private fields. I unfortunately don't have my set up to test multiple crates right now, but I will be *extremely* amused if it turns out you can exhaustively match a `#[non_exhaustive]` struct. Downright dtolnay/rust-quiz material.)

Author: > it could be justified in forbidding matching over any type with inaccessible fields (or #[non_exhaustive]).

This is somewhat deliberate I think: a library can choose to expose certain constants that people can match on despite there being private fields.

I don't think this will let you do exhaustive matching on non_exhaustive enums. For structs there's no separate notion of exhaustiveness anyway, it's all per-field?

Reviewer 5: For structs and struct variants, `#[non_exhaustive]` means "I may add fields in the future so you must always deconstruct with `, ..`". That doesn't change anything for constants-interpreted-as-patterns. I can't find a test for that case though, I'll add one when I get the chance

Reviewer 6: > I will be _extremely_ amused if it turns out you can exhaustively match a `#[non_exhaustive]` struct.

You can :facepalm: 
```rust
# bar/src/lib.rs
#[non_exhaustive]
#[derive(PartialEq, Eq)]
pub struct Bar(bool);
pub const TRUE: Bar = Bar(true);
pub const FALSE: Bar = Bar(false);
```
```rust
# foo/src/main.rs
fn main() {
    match bar::TRUE {
        bar::TRUE => {}
        bar::FALSE => {}
    }
}
```
```console
> cargo check
    Checking bar v0.1.0 (/tmp/scratch.rust.2023-12-17T12-41.GFlTlw/foo/bar)
    Checking foo v0.1.0 (/tmp/scratch.rust.2023-12-17T12-41.GFlTlw/foo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.11s
```

Author: That doesn't seem surprising to me, since this does not deconstruct the value at all, it just compares for equality.

But if this should lint/error then please file an issue.

Reviewer 5: I don't think that should error. `#[non_exhaustive]` means "adding fields musn't be a breaking change", which is the case when using constants. The tricky case is accessing private fields, which is addressed in your RFC.

Reviewer 7: > I don't think that should error. `#[non_exhaustive]` means "adding fields musn't be a breaking change", which is the case when using constants. The tricky case is accessing private fields, which is addressed in your RFC.

How would you add a non-unit type field without breaking the match?
Adding such fields would add more possible states, that would not be coverable by the two existing const patterns.

Reviewer 5: Oh hm fair point. Maybe it should be opt-in like for private fields

Reviewer 8: This appears to be a new (at least to me) semver hazard. I think it'd be very surprising to maintainers that it's possible to cause a breaking change by adding/tweaking a private field on a `#[non_exhaustive]` type.

I think this should either be a future-incompat warning -> error, or at least a lint. If one of the folks here who discovered this issue agrees and would like to open an issue for that, please tag me -- I'd like to keep on top of it for the sake of `cargo-semver-checks`. If nobody is interested in opening that issue, I'd be happy to do it.

Author: Yeah please link the issue from here once it got created. :)

Reviewer 7: As I looks like no has opened an issue yet, I have opened https://github.com/rust-lang/rust/issues/119264

    but only if one of the involved crates implements `StructuralPartialEq` incorrectly.
    This can lead to semver issues and logic bugs, but that is all allowed for safe traits.
    However, this also means unsafe code cannot rely on `==` and pattern matching having the same behavior.
    To make the trait unsafe, the logic for `derive(PartialEq)` should use `#[allow_internal_unsafe]` to still pass `forbid(unsafe_code)`.
  + What should the `StructuralPartialEq` trait be called?
    The current name can be considered confusing because the trait reflects a shallow property, while the value-based check performed when a constant is used in a pattern is defined recursively.
  + Trait bounds `T: StructuralPartialEq` seem like a strange concept. Should we really allow them?

- To avoid interpreting `derive(PartialEq)` as a semver-stable promise of being able to structurally match this type,
  we could introduce an explicit `derive(PartialEq, StructuralPartialEq)`.
  However, that would be a massive breaking change, so it can only be done over an edition boundary.

>   It probably would also want to come with some way to say "derive me all the usual traits" so that one does not have to spell out so many trait names.

Reviewer 9: One potential alternative to this is to allow negative impls for `StructuralPartialEq` for derived `PartialEq`, and/or perhaps to allow typing `derive(PartialEq, !StructuralPartialEq)`.

If we think of `StructuralPartialEq` "like an auto trait" since it's infectious, this should make some amount of sense.

    ```rust
    enum E { Var1, Var2 }

    const BEST_VAR: E = E::Var1;

    fn is_best(e: E) -> bool { matches!(e, BEST_VAR) }
    fn is_var1(e: E) -> bool { matches!(e, E::Var1) }
    ```

    It would be very surprising if those two functions could behave differently.
  It follows that we cannot allow `PartialEq` to affect the behavior of constants in patterns.

- We do not want to expose equality tests on types where the library author did not explicitly expose an equality test.
  This means not allowing matching on constants whose type does not implement `PartialEq`.
  It also means not allowing matching on constants where running `PartialEq` would behave different from the corresponding pattern:
  the pattern can be accessing private fields, and the `PartialEq` implementation provided by the library might be treating those fields in a particular way;
  we should not let people write patterns that "bypass" any such treatment and do structural matching when the crate author does not provide a structural `PartialEq`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Constants can be used as patterns, but only if their type implements `PartialEq`.

> Moreover, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields (recursively):

Reviewer 10: Is one of these "recursively"s redundant?

I'd expect "applies recursively to their fields" to mean fields of fields (recursively), and "applies to their (top-level) fields" to mean just the fields named in the struct.

```suggestion
Moreover, this implementation must be the automatically derived one, and that also applies recursively for the types of their fields:
```

Author: Oh yeah, that's redundant.^^

struct S { f1: i32, f2: E }

const MY_S: S = S { f1: 42, f2: BEST_VAR };

// Removing *either* `derive` or implementing `PartialEq` manually would lead to rejecting this code
fn is_mine(s: S) -> bool { matches!(s, MY_S) }
```

We say that a type that derives `PartialEq` has "structural equality": the
equality of this type is defined fully by the equality on its fields.

For matching on values of enum type, it is sufficient if the actually chosen
variant has structural equality; other variants do not matter:

```rust
struct NonStructuralEq(i32);

impl PartialEq for NonStructuralEq {
    fn eq(&self) -> bool { true }
}

#[derive(PartialEq)]
enum MyEnum { GoodVariant(i32), NonStructuralVariant(NonStructuralEq) }

// This constant *can* be used in a pattern.
const C: MyEnum = MyEnum::GoodVariant(0);
```

This means the eligibility of a constant for a pattern depends on its value, not just on its type.
That is already the case on stable Rust for many years and relied upon by widely-used crates such as [`http`](https://github.com/rust-lang/rust/issues/62411#issuecomment-510604193).

Overall we say that the *value* of the constant must have recursive structural equality,

> which is the case when all the types that actually appear recursively in the type (ignoring "other" enum variants) have structural equality.

Reviewer 10: Is `derive(PartialEq)` necessary on the enum, or just on the contents of the enum variant that's used in the constant?

I think it's required, but saying "appear recursively in the type of the value" is potentially ambiguous. 

```suggestion
which is the case when all the types that actually appear recursively in the type (ignoring "other" enum variants) have structural equality. For a variant to have structural equality, the enum must derive `PartialEq`.
```

Reviewer 10: Ah I see this is explained in the reference section. Perhaps this clarification isn't needed then.

///
/// All that said, this is a safe trait, so violating these requirements
/// can only lead to logic bugs or accidentally exposing an equality test that your
/// library would otherwise not provide, not to unsoundness.
trait StructuralPartialEq: PartialEq {}
```

This trait is automatically implemented when writing `derive(PartialEq)`.
For this RFC to be implemented, the trait can remain unstable can hence cannot be implemented directly by users.
In the future, it might be possible for libraries that implement `PartialEq` by hand (for instance for performance reasons) to also implement `StructuralPartialEq` by hand, if they can promise that the comparison behaves as documented.
(See "Future possibilities" for some of the open questions around that option.)
The trait has `PartialEq` as a supertrait because its entire contract only makes sense for types that implement `PartialEq`.

Range patterns are only allowed on integers, `char`, and floats; for floats, neither end must be a `NaN`.

The *behavior* of such a constant as a pattern is the same as the corresponding native pattern.
On floats are raw pointers, pattern matching behaves like `==`,

> which means in particular that the value `-0.0` matches the pattern `0.0`, and NaN values match no pattern.

Reviewer 10: Is a wildcard considered a pattern?

```suggestion
which means in particular that the value `-0.0` matches the pattern `0.0`, and NaN values only match wildcard patterns (`_`).
```

Author: Good catch, we should mention wildcards.
